domain/lifewatcher.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package domain
+
+import (
+	"context"
+
+	"github.com/juju/collections/set"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/changestream"
+	coredatabase "github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/logger"
+	"github.com/juju/juju/core/watcher/eventsource"
+	"github.com/juju/juju/domain/life"
+)
+
+// LifeGetter is a function which looks up life values of the entities with the specified IDs.
+type LifeGetter func(ctx context.Context, db coredatabase.TxnRunner, ids ...string) (map[string]life.Life, error)
+
+// LifeStringsWatcherMapperFunc returns a namespace watcher mapper function which emits
+// events when the life of an entity changes. The supplied lifeGetter func is used to
+// retrieve a map of life values, keyed on IDs which must match those supplied by the
+// source event stream.
+// The source event stream may supply ids the caller is not interested in. These are
+// filtered out after loading the current values from state.
+func LifeStringsWatcherMapperFunc(logger logger.Logger, lifeGetter LifeGetter) eventsource.Mapper {
+	knownLife := make(map[string]life.Life)
+
+	return func(ctx context.Context, db coredatabase.TxnRunner, changes []changestream.ChangeEvent) (_ []changestream.ChangeEvent, err error) {
+		defer func() {
+			logger.Errorf("running life watcher mapper func: %v", err)
+		}()
+
+		events := make(map[string]changestream.ChangeEvent, len(changes))
+
+		// Extract the ids of the changed entities.
+		ids := set.NewStrings()
+		for _, change := range changes {
+			events[change.Changed()] = change
+			ids.Add(change.Changed())
+		}
+		logger.Debugf("got changes for ids: %v", ids.Values())
+
+		// First record any deleted entities and remove from the
+		// set of ids we are interested in looking up the life for.
+		latest := make(map[string]life.Life)
+		for _, change := range events {
+			if change.Type() == changestream.Delete {
+				latest[change.Changed()] = life.Dead
```

Comment:
We record dead to 2 in the DB. Why do we have to cache the removed to dead?

Reply:
It's so it can re-evaulate later on.

The name "latest" isn't the best here.

Reply:
> It's so it can re-evaulate later on.
> 
> The name "latest" isn't the best here.

Suggestions? It is the "latest" though. The original logic used this name also.

Reply:
If two people have to ask or question if it's correct, I'm going out on a limb, that the name might not be the best.

Reply:
Unsure what it should be, it's just not the _best_.